File: 2802-find-the-punishment-number-of-an-integer/2802-find-the-punishment-number-of-an-integer.cpp
class Solution {
public:
    bool backtrack(string s, int sq, int idx, int sum) {
        if(idx == s.size()){
            return sum == sq;
        }

        int val = 0;
        for (int i = idx; i < s.size(); i++) {
            val = val * 10 + (s[i] - '0');
            if (val + sum > sq) break;
            if (backtrack(s, sq, i + 1, sum + val)) {
                return true; 
            }
        }
        return false;
        
    }

    int punishmentNumber(int n) {
        int res = 0;
        for(int i = 0; i < n + 1; i++) {
            int sq = pow(i, 2);
            string sqstr = to_string(sq);

            if (backtrack(sqstr, i, 0, 0)) res += sq;
        }

        return res;
    }
};

File: 2802-find-the-punishment-number-of-an-integer/2802-find-the-punishment-number-of-an-integer.py
class Solution:
    def punishmentNumber(self, n: int) -> int:
        def partition(i, cur, tar, s):
            if i == len(s):
                return cur == tar

            val = 0
            for j in range(i, len(s)):
                val = val*10 + (ord(s[j]) - ord('0'))
                if partition(j+1, cur + val, tar, s):
                    return True

            return False
        
        res = 0
        for i in range(1,n+1):
            if partition(0, 0, i, str(i*i)):
                res += i**2

        return res